service/storage_service.cc

```diff
@@ -5984,7 +5984,17 @@ future<> storage_service::clone_locally_tablet_storage(locator::global_tablet_id
 // Streams data to the pending tablet replica of a given tablet on this node.
 // The source tablet replica is determined from the current transition info of the tablet.
 future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
-    return do_tablet_operation(tablet, "Streaming", [this, tablet] (locator::tablet_metadata_guard& guard) -> future<> {
+    if (utils::get_local_injector().enter("delay_base_tablet_streaming_by_1s")) {
+        if (!_db.local().find_column_family(tablet.table).schema()->is_view()) {
+            co_await seastar::sleep(std::chrono::seconds(1));
+        }
+    }
+    if (utils::get_local_injector().enter("delay_view_tablet_streaming_by_1s")) {
+        if (_db.local().find_column_family(tablet.table).schema()->is_view()) {
+            co_await seastar::sleep(std::chrono::seconds(1));
+        }
+    }
```

Comment:
@wmitros I think this is not the right place to inject sleep since streaming is bypassed when decreasing RF. I think the right place is handle_tablet_migration() in topology coordinator (inside the streaming case).

Reply:
That's possibly the reason why you were unable to reproduce it with decreasing.

Reply:
If we're bypassing streaming when decreasing RF, how can base and view desync at this step? In `handle_tablet_migration` we wait to advance all tablet migrations: https://github.com/scylladb/scylladb/blob/f69ebc179703265a0709682cdd04fbb0a868f867/service/topology_coordinator.cc#L1120
And I see in the logs that the tablet transition initiated by the `ALTER KEYSPACE` has both base and view tablet migrations:
```
TRACE 2024-11-21 14:38:33,844 [shard 0: gms] raft_topology - do update [{canonical_mutation: table_id fd4f7a46-96bd-3e73-91bf-99eb77e82a5c schema_version 21c69034-ac01-302a-8128-ad4303c8e8c4 partition_key pk{0010a2b09f20a80d11ef9ce1730ea7377103} partition_tombstone {tombstone: none}{row {position: clustered, ckp{00087fffffffffffffff}, 0} tombstone {row_tombstone: none} marker {row_marker: }, column new_replicas atomic_cell{580a63ee-6cbc-4625-88e4-f446db46ec5c:0, 9e330ce2-de89-4472-8867-391221b3d634:0,ts=1732196313844181,expiry=-1,ttl=0}, column stage atomic_cell{allow_write_both_read_old,ts=1732196313844181,expiry=-1,ttl=0}, column transition atomic_cell{rebuild,ts=1732196313844181,expiry=-1,ttl=0}}}, {canonical_mutation: table_id fd4f7a46-96bd-3e73-91bf-99eb77e82a5c schema_version 21c69034-ac01-302a-8128-ad4303c8e8c4 partition_key pk{0010a5b330c0a80d11ef9ce1730ea7377103} partition_tombstone {tombstone: none}{row {position: clustered, ckp{00087fffffffffffffff}, 0} tombstone {row_tombstone: none} marker {row_marker: }, column new_replicas atomic_cell{ddc3f6d4-96dc-40cb-8512-4e8104fc07b3:1, 580a63ee-6cbc-4625-88e4-f446db46ec5c:1,ts=1732196313844181,expiry=-1,ttl=0}, column stage atomic_cell{allow_write_both_read_old,ts=1732196313844181,expiry=-1,ttl=0}, column transition atomic_cell{rebuild,ts=1732196313844181,expiry=-1,ttl=0}}}, {canonical_mutation: table_id 5be1feb3-929e-3df9-8da8-d771c36129a0 schema_version 1599bc92-e8da-3520-8f9f-87e01c8c00f5 partition_key pk{0008746f706f6c6f6779} partition_tombstone {tombstone: none}, static column global_topology_request atomic_cell{DEAD,ts=1732196313844181,deletion_time=1732196313}, static column global_topology_request_id atomic_cell{DEAD,ts=1732196313844181,deletion_time=1732196313}, static column transition_state atomic_cell{tablet migration,ts=1732196313844181,expiry=-1,ttl=0}, static column version atomic_cell{66,ts=1732196313844181,expiry=-1,ttl=0}}, {canonical_mutation: table_id b3068525-d0f2-39c2-8250-584f9b1bf460 schema_version 4d10285c-c6bf-3513-b21c-f7ade15baf5e partition_key pk{0010e72c0324a80d11ef57c59225f11797e4} partition_tombstone {tombstone: none}{row {position: clustered, ckp{}, 0} tombstone {row_tombstone: none} marker {row_marker: 1732196313823722 2629746 1734826059}, column done atomic_cell{true,ts=1732196313823722,expiry=1734826059,ttl=2629746}, column end_time atomic_cell{2024-11-21T13:38:33.844Z,ts=1732196313823722,expiry=1734826059,ttl=2629746}, column error atomic_cell{,ts=1732196313823722,expiry=1734826059,ttl=2629746}}}, {canonical_mutation: table_id fa0ea2bd-608f-3e74-9b1e-b84b46b33adf schema_version 73c8abde-708f-3e5c-a1b5-52c823cb7b33 partition_key pk{00026b73} partition_tombstone {tombstone: none}}, {canonical_mutation: table_id abac5682-dea6-31c5-b535-b3d6cffd0fb6 schema_version e79ca8ba-6556-3f7d-925a-7f20cf57938c partition_key pk{00026b73} partition_tombstone {tombstone: none}{row {position: clustered, ckp{}, 0} tombstone {row_tombstone: none} marker {row_marker: }, column durable_writes atomic_cell{true,ts=1732196313844181,expiry=-1,ttl=0}, column replication atomic_cell{({class : org.apache.cassandra.locator.NetworkTopologyStrategy}, {datacenter1 : 2}),ts=1732196313844181,expiry=-1,ttl=0}}}] reason ALTER tablets KEYSPACE called with options: {"replication:datacenter1": "2", "replication:class": "NetworkTopologyStrategy", "durable_writes:durable_writes": "true"}
```
For comparison, here a log after decreasing RF where the materialized view is not present (we set only 1 tablet to `allow_write_both_read_old` instead of 2):
```
TRACE 2024-11-21 14:49:25,759 [shard 0: gms] raft_topology - do update [{canonical_mutation: table_id fd4f7a46-96bd-3e73-91bf-99eb77e82a5c schema_version 21c69034-ac01-302a-8128-ad4303c8e8c4 partition_key pk{0010a2b09f20a80d11ef9ce1730ea7377103} partition_tombstone {tombstone: none}{row {position: clustered, ckp{00087fffffffffffffff}, 0} tombstone {row_tombstone: none} marker {row_marker: }, column new_replicas atomic_cell{580a63ee-6cbc-4625-88e4-f446db46ec5c:0, 9e330ce2-de89-4472-8867-391221b3d634:0,ts=1732196965758842,expiry=-1,ttl=0}, column stage atomic_cell{allow_write_both_read_old,ts=1732196965758842,expiry=-1,ttl=0}, column transition atomic_cell{rebuild,ts=1732196965758842,expiry=-1,ttl=0}}}, {canonical_mutation: table_id 5be1feb3-929e-3df9-8da8-d771c36129a0 schema_version 1599bc92-e8da-3520-8f9f-87e01c8c00f5 partition_key pk{0008746f706f6c6f6779} partition_tombstone {tombstone: none}, static column global_topology_request atomic_cell{DEAD,ts=1732196965758842,deletion_time=1732196965}, static column global_topology_request_id atomic_cell{DEAD,ts=1732196965758842,deletion_time=1732196965}, static column transition_state atomic_cell{tablet migration,ts=1732196965758842,expiry=-1,ttl=0}, static column version atomic_cell{113,ts=1732196965758842,expiry=-1,ttl=0}}, {canonical_mutation: table_id b3068525-d0f2-39c2-8250-584f9b1bf460 schema_version 4d10285c-c6bf-3513-b21c-f7ade15baf5e partition_key pk{00106bbb9086a80f11ef51864706f9221652} partition_tombstone {tombstone: none}{row {position: clustered, ckp{}, 0} tombstone {row_tombstone: none} marker {row_marker: 1732196965720487 2629746 1734826711}, column done atomic_cell{true,ts=1732196965720487,expiry=1734826711,ttl=2629746}, column end_time atomic_cell{2024-11-21T13:49:25.758Z,ts=1732196965720487,expiry=1734826711,ttl=2629746}, column error atomic_cell{,ts=1732196965720487,expiry=1734826711,ttl=2629746}}}, {canonical_mutation: table_id fa0ea2bd-608f-3e74-9b1e-b84b46b33adf schema_version 73c8abde-708f-3e5c-a1b5-52c823cb7b33 partition_key pk{00026b73} partition_tombstone {tombstone: none}}, {canonical_mutation: table_id abac5682-dea6-31c5-b535-b3d6cffd0fb6 schema_version e79ca8ba-6556-3f7d-925a-7f20cf57938c partition_key pk{00026b73} partition_tombstone {tombstone: none}{row {position: clustered, ckp{}, 0} tombstone {row_tombstone: none} marker {row_marker: }, column durable_writes atomic_cell{true,ts=1732196965758842,expiry=-1,ttl=0}, column replication atomic_cell{({class : org.apache.cassandra.locator.NetworkTopologyStrategy}, {datacenter1 : 2}),ts=1732196965758842,expiry=-1,ttl=0}}}] reason ALTER tablets KEYSPACE called with options: {"replication:datacenter1": "2", "replication:class": "NetworkTopologyStrategy", "durable_writes:durable_writes": "true"}
```

Reply:
I agree, however there are multiple stages (streaming, cleanup) in which a migration can lose the race to another one (even if they happened to start at same time), and we need to make it robust thinking forward, so a future change in the handler won't make the potential problem reproducible in production. But I also agree it decreases the severity.

Reply:
It's good that you reproduced it with increasing RF. so let's stick to it as our reproducer.